Draft: `VRMC_springBone_limit`

0b5vr opened the pull request:
This is the first draft of `VRMC_springBone_limit`.

Added the Japanese spec and the schema, English spec is just a placeholder.

0b5vr committed:
feat: The first draft of `VRMC_springBone_limit`

Added the Japanese spec and the schema, English spec is just a placeholder

0b5vr committed:
feat: `VRMC_springBone_limit`, swap phi and theta

phi ranges from -PI to PI, theta ranges from -PI/2 to PI/2

0b5vr committed:
feat: `VRMC_springBone_limit`, update details of rotation and application order

0b5vr committed:
feat: `VRMC_springBone_limit`, fix example JSON

0b5vr committed:
docs: VRMC_springBone_limit, remove a TODO

it's gonna be radians instead of degrees

0b5vr committed:
feat: `VRMC_springBone_limit`, add reference implementation

ousttrue reviewed (dismissed):
LGTM

ousttrue commented on specification/VRMC_springBone_limit-1.0/README.ja.md:
```diff
@@ -0,0 +1,471 @@
+
+以下に、本拡張で定義する各リミットの参考実装を示します。
+
+以下の参考実装において `tailDir` は、制限するjointの向きです。 `tailDir` は正規化されているものとします。
```
tailDir が突然出てくる感じがします。
node.rotation もしくは node.localRotation から話が始まるべきと思います。
もしくは既存の springBone 計算の中間値を使っているなら
それが出てくる過程も書いておく方がよいと思います。
要するに、コピペしてから使うところまでが遠いです。

ousttrue commented on specification/VRMC_springBone_limit-1.0/README.ja.md:
````diff
@@ -0,0 +1,471 @@
+
+```ts
+// Y+方向からjointのheadからtailに向かうベクトルへの最小回転
+let axisRotation = fromToQuaternion(vec3(0, 1, 0), boneAxis);
````
boneAxis の意味を類推する必要があって、
難解になっているような。

0b5vr replied:
フィードバックありがとうございます。 `boneAxis` については、 `VRMC_springBone` 側の擬似コードでも定義されている「そのJointが対象とする子Nodeの、ローカル空間におけるレスト状態の伸びる方向」ですが、改めてこちらの文書でも説明することとします。

ousttrue commented on specification/VRMC_springBone_limit-1.0/README.ja.md:
```diff
@@ -0,0 +1,471 @@
+tailDir = tailDir.normalized;
+
+// tailDirのy要素をjointに設定されたangleの余弦と比較する
+let cosAngle = cos(joint.angle);
```
joint => limitParam など limit の設定値であることが見た目に分かる名前の方がわかりやすいと思われます。後続も同様です。

0b5vr committed:
docs: `VRMC_springBone_limit`, update `tailDir` description for clarity and provide pseudo-code example

Suggested by @ousttrue: https://github.com/vrm-c/vrm-specification/pull/496#discussion_r2209331291

0b5vr committed:
docs: `VRMC_springBone_limit`, add explanation for `boneAxis` in the pseudocode

Suggested by @ousttrue: https://github.com/vrm-c/vrm-specification/pull/496#discussion_r2209357727

0b5vr committed:
docs: `VRMC_springBone_limit`, update pseudocode to reference `limit` instead of `joint` for limit params

Suggested by @ousttrue: https://github.com/vrm-c/vrm-specification/pull/496#discussion_r2209395749

ousttrue commented:
対応ありがとうございます。
読みやすくなりました 👍

0b5vr committed:
docs: `VRMC_springBone_limit` clarify the timing to apply limits in the reference implementations section

ousttrue commented:
実装してみたところ、各ロジックについて Json に記述された角度を 0.5 して適用すると Gizmo とぴったりになるようです。
設定 UI では扇形の両端の角度を設定して Json に記述しますが、計算上は tangent するので半分しか要らないという解釈問題が
ありそうです。

0b5vr committed:
docs: `VRMC_springBone_limit`, clarify the meaning of parameter angles of spherical limit

While cone limit and hinge limit explicitly define that the parameter angles are the maximum angles allowed from the reference direction, spherical limit were vague about it

0b5vr committed:
feat: NORMATIVE CHANGE to draft springBone limit spec, change spherical limit angles to pitch-yaw instead of phi-theta

phi-theta can be confusing since there are multiple conventions; we instead simply follow the well-used notation

ousttrue commented on specification/VRMC_springBone_limit-1.0/figures/spherical-limit-pitch-yaw.png:
👍 limit 空間の x軸回転が pitch

ousttrue reviewed (dismissed):
👍

0b5vr committed:
feat: NORMATIVE CHANGE to draft springBone limit spec, make it rotate 180deg around x axis when the boneAxis is exactly (0, -1, 0) since the minimal rotation is indeterminate

also update the pseudocode to more primitive way and that reflects the spec update

0b5vr committed:
fix: springBone_limit, Fix pseudocode of rotation

the cross production was incorrect
(0, 1, 0) × (x, y, z) = (z, 0, -x)

0b5vr committed:
fix: springBone_limit, Respect angle bounds in pseudocode

Clamp cone and hinge angles to [0, pi], spherical pitch to [0, pi],
and spherical yaw to [0, pi/2] before applying the limits.

This keeps the reference implementation consistent with the documented
lower and upper bounds.

0b5vr committed:
feat: NORMATIVE CHANGE to draft springBone limit spec, define fallback directions for singular cases

Limit projections can be indeterminate when the tail points along axes
that lose the directional component required by a limit.

Define deterministic local-space fallback directions for cone, hinge,
and spherical limits, and update the reference pseudocode to avoid zero
divisions and undefined angles.

Keep the pseudocode mathematically exact while recommending that
implementations apply precision-appropriate floating-point safeguards.

0b5vr committed:
feat: NORMATIVE CHANGE to draft springBone limit spec, quaternion properties must be normalized

0b5vr committed:
docs: `VRMC_springBone_limit`, clarify optional limit application

Limit evaluation may be reduced or disabled depending on the target
platform, runtime settings, or LOD requirements.

Clarify that the recommended application timings apply when limit
evaluation is enabled, and note that the final tail direction is not
guaranteed to remain within the limit when evaluation is reduced or
disabled.

0b5vr committed:
feat: NORMATIVE CHANGE to draft springBone limit spec, ignore limits on terminal joints

The terminal joint in a spring chain acts only as the tail, so a limit
attached to it cannot affect the simulation.

Require implementations to ignore the extension on terminal joints and
exporters to omit it from those joints.

0b5vr committed:
docs: `VRMC_springBone_limit`, clarify required values

The specVersion and exactly-one limit requirements were already
expressed in the property descriptions and JSON Schema, but their
normative strength was not explicit in the text.

Mark both requirements as MUST to make conformance expectations
unambiguous without changing the schema.

0b5vr committed:
docs: spring bone limit, make RFC 2119 keywords bold and clarify angle constraints

Normative requirements were not consistently emphasized, and the angle
constraints did not state their RFC 2119 strength explicitly.

Bold all RFC 2119 keywords and mark the non-negative angle requirements
and upper-bound interpretation rules as MUST for cone, hinge, and
spherical limits.

0b5vr committed:
docs: springbone limit, trivial, fix comment inconsistency in pseudocode

0b5vr commented:
英訳の追加 https://github.com/vrm-c/vrm-specification/issues/514 ならびに文意の改善を行いました。Normative changesは含まれません。
